src/python/pants/backend/go/build.py

```diff
@@ -0,0 +1,354 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import logging
+import os
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Dict, List, Tuple
+
+from pants.backend.go.distribution import GoLangDistribution
+from pants.backend.go.target_types import GoBinaryMainAddress, GoBinaryName, GoImportPath, GoSources
+from pants.build_graph.address import Address, AddressInput
+from pants.core.goals.package import (
+    BuiltPackage,
+    BuiltPackageArtifact,
+    OutputPathField,
+    PackageFieldSet,
+)
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, FileContent, MergeDigests, Snapshot
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.target import (
+    FieldSet,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+    WrappedTarget,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+_logger = logging.getLogger(__name__)
+
+
+class GoBuildSubsystem(GoalSubsystem):
+    name = "go-build"
+    help = """Compile Go targets that contain source code (i.e., `go_package`)."""
+
+
+class GoBuildGoal(Goal):
+    subsystem_cls = GoBuildSubsystem
+
+
+@dataclass(frozen=True)
+class BuildGoPackageFieldSet(FieldSet):
+    required_fields = (GoSources, GoImportPath)
+    sources: GoSources
+    import_path: GoImportPath
+
+
+@dataclass(frozen=True)
+class BuildGoPackageRequest:
+    field_sets: BuildGoPackageFieldSet
+    is_main: bool = False
+
+
+@dataclass(frozen=True)
+class BuiltGoPackage:
+    import_path: str
+    object_digest: Digest
+
+
+@dataclass(frozen=True)
+class GoBinaryFieldSet(PackageFieldSet):
+    required_fields = (GoBinaryName, GoBinaryMainAddress)
+
+    binary_name: GoBinaryName
+    main_address: GoBinaryMainAddress
+    output_path: OutputPathField
+
+
+@dataclass(frozen=True)
+class EnrichedGoLangDistribution:
+    stdlib_packages: Dict[str, str]
+
+
+@dataclass(frozen=True)
+class EnrichGoLangDistributionRequest:
+    downloaded_goroot: DownloadedExternalTool
+
+
+@goal_rule
+async def run_go_build(targets: Targets) -> GoBuildGoal:
+    gets_for_targets = [
+        Get(BuiltGoPackage, BuildGoPackageRequest(BuildGoPackageFieldSet.create(target)))
+        for target in targets
+        if BuildGoPackageFieldSet.is_applicable(target)
+    ]
+    _ = await MultiGet(gets_for_targets)
```

Comment:
Imo you can inline the above comprehension. For one, using a generator instead of list comprehension will save on creating an intermediate collection.

```suggestion
    await MultiGet(gets_for_targets)
```